kic_module.c

```diff
@@ -0,0 +1,183 @@
+                kic_reset();
+                pr_info("KIC: Counter has been reset.\n");
+                break;
+        }// case closed
+        case KIC_GET_COUNT_QUERY: {     // option for retrieving count
+                pr_info("KIC: Count retrieved at %u.\n", kic_interrupt_count);
+                put_user(kic_interrupt_count, (unsigned int __user *)ioctl_carry);
+                break;
+        }// case closed
+        case KIC_GET_TIME_QUERY: {      // option for retrieving time and date
+                if (kic_time == 0) {
+                        pr_alert("KIC: Timer has not yet been reset.\n");
+                        put_user(-1, (long int __user *)ioctl_carry);
+                        break;
+                }
+                put_user(kic_time, (long int __user *)ioctl_carry);
+                pr_info("KIC: Time retrieved.\n");
+                break;
+        }// case closed
+        }// switch closed
+        atomic_set (&already_open, CDEV_NOT_USED);
+        return 0;
+}
+
+static struct file_operations fops = {
+        .owner = THIS_MODULE,
+        .open = kic_device_open,
+        .release = kic_device_release,
+        .unlocked_ioctl = kic_device_ioctl,
+};
+
+        // purpose specific functions //
+// increases count upon detecting a valid interrupt
+static irqreturn_t kic_increment(int irq, void *dev_id)
+{
+        kic_interrupt_count++;
+        return IRQ_HANDLED;
+}
+
+// resets counter and saves time data when invoked
+static void kic_reset(void)
+{
+        kic_time = ktime_get_real();
+        kic_interrupt_count = 0;
+}
+
+        // init and exit //
+static int __init kic_init(void)
+{
+        int irq_ret;
+
+        irq_ret = register_chrdev(KIC_MAJOR, KIC_DEVICE_NAME, &fops);
+        if (irq_ret < 0) {
+                pr_alert("KIC: Registering device failed with error code: %d\n",
+                         irq_ret);
+                goto return_status;
+        }
+        pr_info("KIC: Device assigned major number %d\n", KIC_MAJOR);
+
+
+        kic_cls = class_create(THIS_MODULE, KIC_DEVICE_NAME);
+        if ((void *)kic_cls == ERR_PTR) {
+                pr_alert("KIC: Class creation failed.\n");
+                goto class_fail;
+        }
+
+        /*
+         * may be slightly confusing, casting required for comparison,
+         * pointer returned from device_crate is useful only for error handling
+         */
+        if ((void *)(device_create(kic_cls, NULL, MKDEV(KIC_MAJOR, 0), NULL,
+            KIC_DEVICE_NAME)) == ERR_PTR) {
+                pr_alert("KIC: Creating device failed with error code: %d\n",
+                         irq_ret);
+                goto device_fail;
+        }
+        pr_info("KIC: Device created on /dev/%s\n", KIC_DEVICE_NAME);
+
+
+        irq_ret = request_irq(KEYBOARD_IRQ, kic_increment, KIC_IRQ_FLAGS,
+                                  KIC_DEVICE_NAME, (void *)kic_cls);
+        if (irq_ret < 0) {
```

Comment:
I think you used 8 spaces instead of tab in a lot of code, please fix that.

Reply:
done